li-protobuf/src/main/java/com/linkedin/data/protobuf/OutputStreamWriter.java

```diff
@@ -0,0 +1,359 @@
+    _buffer[_position++] = (byte) (value & 0xFF);
+    _buffer[_position++] = (byte) ((value >> 8) & 0xFF);
+    _buffer[_position++] = (byte) ((value >> 16) & 0xFF);
+    _buffer[_position++] = (byte) ((value >> 24) & 0xFF);
+  }
+
+  /**
+   * Write a variable length 32-bit signed integer.
+   */
+  public final void writeInt32(final int value) throws IOException
+  {
+    if (value >= 0)
+    {
+      writeUInt32(value);
+    }
+    else
+    {
+      // Must sign-extend.
+      writeUInt64(value);
+    }
+  }
+
+  /**
+   * Write a fixed length 64-bit signed integer.
+   */
+  public final void writeFixedInt64(final long value) throws IOException
+  {
+    flushIfNotAvailable(ProtoWriter.FIXED64_SIZE);
+    _buffer[_position++] = (byte) ((int) (value) & 0xFF);
+    _buffer[_position++] = (byte) ((int) (value >> 8) & 0xFF);
+    _buffer[_position++] = (byte) ((int) (value >> 16) & 0xFF);
+    _buffer[_position++] = (byte) ((int) (value >> 24) & 0xFF);
+    _buffer[_position++] = (byte) ((int) (value >> 32) & 0xFF);
+    _buffer[_position++] = (byte) ((int) (value >> 40) & 0xFF);
+    _buffer[_position++] = (byte) ((int) (value >> 48) & 0xFF);
+    _buffer[_position++] = (byte) ((int) (value >> 56) & 0xFF);
+  }
+
+  /**
+   * Write a variable length 64-bit signed integer.
+   */
+  public final void writeInt64(final long value) throws IOException
+  {
+    writeUInt64(value);
+  }
+
+  private void buffer(byte value) throws IOException
+  {
+    _buffer[_position++] = value;
+  }
+
+  /**
+   * Flush any buffered data to the underlying outputstream.
+   */
+  public void flush() throws IOException
+  {
+    _out.write(_buffer, 0, _position);
+    _position = 0;
+  }
+
+  private void flushIfNotAvailable(int requiredSize) throws IOException
+  {
+    if (_limit - _position < requiredSize)
+    {
+      flush();
+    }
+  }
+
+  /**
+   * Write a variable length 32-bit unsigned integer.
+   */
+  public void writeUInt32(int value) throws IOException
+  {
+    flushIfNotAvailable(ProtoWriter.MAX_VARINT32_SIZE);
+    bufferUInt32(value);
+  }
+
+  private void bufferUInt32(int value) throws IOException
+  {
+    while (true)
+    {
+      if ((value & ~0x7F) == 0)
+      {
+        _buffer[_position++] = (byte) value;
+        return;
+      }
+      else
+      {
+        _buffer[_position++] = (byte) ((value & 0x7F) | 0x80);
+        value >>>= 7;
+      }
+    }
+  }
+
+  /**
+   * Write a variable length 64-bit unsigned integer.
+   */
+  public void writeUInt64(long value) throws IOException
+  {
+    flushIfNotAvailable(ProtoWriter.MAX_VARINT64_SIZE);
+    bufferUInt64(value);
+  }
+
+  private void bufferUInt64(long value) throws IOException
+  {
+    while (true)
+    {
+      if ((value & ~0x7FL) == 0)
+      {
+        _buffer[_position++] = (byte) value;
+        return;
+      }
+      else
+      {
+        _buffer[_position++] = (byte) (((int) value & 0x7F) | 0x80);
+        value >>>= 7;
```

Comment:
If we are trying to copy individual bytes to the _buffer, I think we should shift 8?
0b100000000 = 256
256 >> 8 = 1 seems to be right for this case?

Reply:
See how varints are encoded.
https://developers.google.com/protocol-buffers/docs/encoding#varints